core/blockchain.go

```diff
@@ -2617,9 +2619,32 @@ func (bc *BlockChain) UpdateValidatorUptime(slots shard.SlotList, mask *bls.Mask
 				return err
 			}
 
-			if enabled {
-				stats.NumBlocksSigned.Add(stats.NumBlocksSigned, big.NewInt(1))
+			if one := big.NewInt(1); enabled {
+				stats.NumBlocksSigned.Add(stats.NumBlocksSigned, one)
+				stats.NumBlocksMissed.Set(big.NewInt(0))
+			} else {
+				stats.NumBlocksMissed.Add(stats.NumBlocksMissed, one)
 			}
+
+			if stats.NumBlocksMissed.Cmp(slash.MissedThresholdForInactive) == 0 {
+				active, err := bc.ReadActiveValidatorList()
+				if err != nil {
+					return err
+				}
+
+				faithful := []common.Address{}
+				mia := addr.Bytes()
+				for i := range active {
+					if bytes.Compare(mia, active[i].Bytes()) != 0 {
+						faithful = append(faithful, active[i])
+					}
+				}
+				if err := bc.WriteActiveValidatorList(faithful); err != nil {
```

Comment:
This is not the way. ActiveValidatorList is the list of validators that's selected in the committee. By removing it from the list doesn't rule them out from next epoch committee selection. Instead, you should flip the IsActive flag of the validator information in the account state. And when you compute committee in next epoch. don't select the ones with IsActive==false.

Reply:
Since IsActive is part of the account state so it affect state root. So flipping the IsActive field should happen before the state root is calculated. Somewhere in the Finalize() method should work. Similar to the delegation withdrawal payout code.

Reply:
1) I thought it was probably not the right way but was irresistibly easy, will amend. 
2) Also thought that issue would happen but wasn't sure right place, will amend